File: Chapter10/C#/Project/AdvancedMedia/1-AdvanceMedia/AdvMediaPlayer.Web/EncoderMedia/MediaPlayer/PlayList.cs
// <copyright file="PlaylistCollection.cs" company="Microsoft">
//     Copyright © Microsoft Corporation. All rights reserved.
// </copyright>
// <summary>Implements the PlaylistCollection class</summary>
// <author>Microsoft Expression Encoder Team</author>
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Diagnostics;
using System.Windows;
using System.Windows.Browser;
using System.Windows.Media;
using System.Threading;
using System.Xml;
using System.IO.IsolatedStorage;
using System.IO;
using System.Globalization;
using System.Text;

namespace ExpressionMediaPlayer
{
    /// <summary>
    /// exception for invalid playlist
    /// </summary>
    public class InvalidPlaylistException : Exception
    {
        public InvalidPlaylistException() 
            : base(Resources.errorInvalidPlaylist)
        {
        }
        public InvalidPlaylistException(string message)
            : base(message)
        {
        }
        public InvalidPlaylistException(string nodeFound, string nodeExpected)
            : base(String.Format(CultureInfo.CurrentUICulture, Resources.errorInvalidPlaylistNode, nodeFound, nodeExpected))
        {
        }
    }

    // Summary:
    //     Describes how content is resized to fill its allocated space -- taking into account non-square pixels
    public enum StretchNonSquarePixels
    {
        // Summary:
        //     The content preserves its original size and frame aspect ratio
        NoStretch = 0,
        //
        // Summary:
        //     The content is resized to fill the destination dimensions. The frame aspect ratio is preserved.
        StretchToFill = 1,
        //
        // Summary:
        //     The content is resized to fill the destination dimensions. Original size and frame aspect ratio are ignore.
        StretchDistorted = 2,
    }

    [ScriptableType]
    public partial class Playlist : DependencyObject
    {
        #region PlaylistCore
        /// <summary>
        /// playlist options
        /// </summary>
        private bool m_autoLoad;
        private bool m_autoPlay;
        private bool m_autoRepeat;
        private bool m_displayTimeCode;
        private bool m_enableCachedComposition;
        private bool m_enableCaptions;
        private bool m_enableOffline;
        private bool m_enablePopOut;
        private bool m_startMuted;
        private bool m_startWithPlaylistShowing;
        private StretchNonSquarePixels m_stretchMode;
        private Color m_background;

        /// <summary>
        /// The number of items in the playlist that are DRM protected
        /// </summary>
        private int m_countOfDRMItems;
        /// <summary>
        /// The number of items in the playlist that are Live Items;
        /// </summary>
        private int m_countOfLiveItems;
        /// <summary>
        /// The number of items in the playlist that are adaptive Items;
        /// </summary>
        private int m_countOfAdaptiveItems;

        /// <summary>
        /// list of playlist items
        /// </summary>        
        private PlaylistCollection _Items = new PlaylistCollection();
        /// <summary>
        /// playlist changed event
        /// </summary>
        [ScriptableMember]
        public event RoutedEventHandler PlaylistChanged;
        /// <summary>
        /// should playlist cue up 
        /// </summary>
        [Description("Automatically cue video when page is loaded"), DefaultValue(true)]
        public bool AutoLoad 
        { 
            get
            {
                return m_autoLoad;
            }
            set
            {
                m_autoLoad = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// should playlist auto start?
        /// </summary>
        [Description("Automatically start video when cued"), DefaultValue(true)]
        public bool AutoPlay
        {
            get
            {
                return m_autoPlay;
            }
            set
            {
                m_autoPlay = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// should playlist auto repeat?
        /// </summary>
        [Description("Automatically restart video when at end"), DefaultValue(false)]
        public bool AutoRepeat
        {
            get
            {
                return m_autoRepeat;
            }
            set
            {
                m_autoRepeat = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// display timecodes
        /// </summary>
        [Description("Display Timecode")]
        public bool DisplayTimeCode
        {
            get
            {
                return m_displayTimeCode;
            }
            set
            {
                m_displayTimeCode = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// cached composition enabled
        /// </summary>
        [Description("Enable Cached Composition"), DefaultValue(true)]
        public bool EnableCachedComposition
        {
            get
            {
                return m_enableCachedComposition;
            }
            set
            {
                m_enableCachedComposition = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// captions enabled
        /// </summary>
        [Description("Allow closed captions to show"), DefaultValue(true)]
        public bool EnableCaptions
        {
            get
            {
                return m_enableCaptions;
            }
            set
            {
                m_enableCaptions = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// offlining enabled
        /// </summary>
        [Description("Enable Player to be run offline"), DefaultValue(true)]
        public bool EnableOffline
        {
            get
            {
                return m_enableOffline && (m_countOfDRMItems < 1) && (m_countOfLiveItems < 1);
            }
            set
            {
                m_enableOffline = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// popout enabled
        /// </summary>
        [Description("Enable Player popout"), DefaultValue(true)]
        public bool EnablePopOut
        {
            get
            {
                return m_enablePopOut;
            }
            set
            {
                m_enablePopOut = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }

        /// <summary>
        /// HasAdaptiveItems
        /// </summary>
        [Description("Indicated whether there are any Smooth Streaming items in the playlist")]
        public bool HasAdaptiveItems
        {
            get
            {
                return m_countOfAdaptiveItems > 0;
            }
        }


        /// <summary>
        /// start in muted state
        /// </summary>
        [Description("Mute player on start")]
        public bool StartMuted
        {
            get
            {
                return m_startMuted;
            }
            set
            {
                m_startMuted = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }

        /// <summary>
        /// start with playlist visible
        /// </summary>
        [Description("Show playlist on start")]
        public bool StartWithPlaylistShowing
        {
            get
            {
                return m_startWithPlaylistShowing;
            }
            set
            {
                m_startWithPlaylistShowing = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// type of video stretch
        /// </summary>
        [Description("Stretch Mode")]
        public StretchNonSquarePixels StretchNonSquarePixels
        {
            get
            {
                return m_stretchMode;
            }
            set
            {
                m_stretchMode = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// color of background
        /// </summary>
        public Color Background
        {
            get
            {
                return m_background;
            }
            set
            {
                m_background = value;
                if (PlaylistChanged != null)
                {
                    PlaylistChanged(this, null);
                }
            }
        }
        /// <summary>
        /// list of playlist items.
        /// </summary>
        [Description("Playlist")]
        public PlaylistCollection Items
        {
            get
            {
                return _Items;
            }
            set
            {
                _Items = value;
            }
        }
        /// <summary>
        /// init structure
        /// </summary>
        private void Init()
        {
            AutoLoad = true;
            AutoPlay = true;
            AutoRepeat = false;
            EnableCachedComposition = true;
            EnableCaptions = true;
            EnablePopOut = true;
            EnableOffline = true;
            m_countOfDRMItems = 1;
            m_countOfLiveItems = 1;
            m_countOfAdaptiveItems = 0;
        }
        /// <summary>
        /// construct a playlistitem, provided for scripting.
        /// </summary>
        /// <returns>new playlist item</returns> 
        public PlaylistItem CreateNewPlaylistItem()
        {
            return new PlaylistItem(Items);
        }
        /// <summary>
        /// playlist constructor
        /// </summary>
        public Playlist()
        {
            this.StretchNonSquarePixels = StretchNonSquarePixels.NoStretch;
            _Items = new PlaylistCollection();
            _Items.CollectionChanged += new System.Collections.Specialized.NotifyCollectionChangedEventHandler(Items_CollectionChanged);
            Init();         
        }
        /// <summary>
        /// inform playlist changed if the items collection changed
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void Items_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            if (PlaylistChanged!=null)
            {
                PlaylistChanged(this, null);
            }
            m_countOfDRMItems = 0;
            m_countOfLiveItems = 0;
            m_countOfAdaptiveItems = 0;
            foreach (var item in this.Items)
            {
                if (item.DRM)
                {
                    m_countOfDRMItems++;
                }
                if (item.IsAdaptiveStreaming)
                {
                    m_countOfAdaptiveItems++;
                    string source = item.MediaSource.OriginalString.ToLower();
                    if (source.Contains(@".isml/manifest"))
                    {
                        m_countOfLiveItems++;
                    }
                }
                else
                {
                    if (item.MediaSource.IsAbsoluteUri)
                    {
                        string scheme = item.MediaSource.Scheme.ToLower();
                        if (scheme == ("mms:"))
                        {
                            m_countOfLiveItems++;
                        }
                    }
                }
            }
        }
        /// <summary>
        /// create playlist from XML string
        /// </summary>
        /// <param name="playlistXml">XML string</param>
        /// <returns>playlist</returns>
        public static Playlist Create(string playlistXml)
        {
            Playlist playlist = new Playlist();
            playlist.ParseXml(playlistXml);
            return playlist;
        }
        /// <summary>
        /// parse playlist from XML
        /// </summary>
        /// <param name="playlistXml"></param>
        public void ParseXml(string playlistXml)
        {
            UTF8Encoding enc = new UTF8Encoding();
            using (MemoryStream ms = new MemoryStream(enc.GetBytes(playlistXml)))
            {
                XmlReaderSettings xmlrs = new XmlReaderSettings();
                xmlrs.IgnoreComments = true;
                xmlrs.IgnoreWhitespace = true;
                XmlReader reader = XmlReader.Create(ms, xmlrs);
                Deserialize(reader);
            }
            if (PlaylistChanged != null)
            {
                PlaylistChanged(this, null);
            }
        }
        #endregion

        #region Serialization
        /// <summary>
        /// top level XML node for this class
        /// </summary>
        internal const string xmlNode = "Playlist";
        /// <summary>
        /// deserialise this object
        /// </summary>
        /// <param name="reader">XmlReader to deserialize from</param>
        /// <returns>this</returns>
        internal void Deserialize(XmlReader reader)
        {
            if (!reader.IsStartElement(Playlist.xmlNode))
                throw new InvalidPlaylistException();

            Init();
            reader.Read();
            while (!(reader.Name == Playlist.xmlNode && reader.NodeType == XmlNodeType.EndElement))
            {
                if (reader.IsStartElement("AutoLoad"))
                    this.AutoLoad = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("AutoPlay"))
                    this.AutoPlay = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("AutoRepeat"))
                    this.AutoRepeat = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("DisplayTimeCode"))
                    this.DisplayTimeCode = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("EnableCachedComposition"))
                    this.EnableCachedComposition = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("EnableCaptions"))
                    this.EnableCaptions = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("EnablePopOut"))
                    this.EnablePopOut = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("EnableOffline"))
                    this.EnableOffline = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("StartMuted"))
                    this.StartMuted = reader.ReadElementContentAsBoolean();
                else if (reader.IsStartElement("StartWithPlaylistShowing"))
                {
                    this.StartWithPlaylistShowing = reader.ReadElementContentAsBoolean();
                }
                else if (reader.IsStartElement("StretchMode"))
                {
                    Stretch tmp = (Stretch)Enum.Parse(typeof(Stretch), reader.ReadElementContentAsString(), false);
                    switch(tmp)
                    {
                        case Stretch.None:
                            this.StretchNonSquarePixels = StretchNonSquarePixels.NoStretch;
                            break;
                        case Stretch.Uniform:
                        case Stretch.UniformToFill:
                            this.StretchNonSquarePixels = StretchNonSquarePixels.StretchToFill;
                            break;
                        case Stretch.Fill:
                            this.StretchNonSquarePixels = StretchNonSquarePixels.StretchDistorted;
                            break;
                    }
                }
                else if (reader.IsStartElement("StretchNonSquarePixels"))
                {
                    StretchNonSquarePixels tmp = (StretchNonSquarePixels)Enum.Parse(typeof(StretchNonSquarePixels), reader.ReadElementContentAsString(), false);
                    this.StretchNonSquarePixels = tmp;
                }
                else if (reader.IsStartElement(PlaylistCollection.xmlNode))
                {
                    this.Items.Clear();
                    this.Items.Deserialize(reader);
                }
                else if (reader.IsStartElement())
                    throw new InvalidPlaylistException(reader.Name, Playlist.xmlNode);
                else if (!reader.Read())
                    break;
            }
        }
        /// <summary>
        /// serialize this object
        /// </summary>
        /// <param name="writer">XmlWriter</param>
        internal void Serialize(XmlWriter writer)
        {
            writer.WriteStartElement(Playlist.xmlNode);
            writer.WriteElementString("AutoLoad", this.AutoLoad.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("AutoPlay", this.AutoPlay.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("AutoRepeat", this.AutoRepeat.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("DisplayTimeCode", this.DisplayTimeCode.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("EnableCachedComposition", this.EnableCachedComposition.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("EnableCaptions", this.EnableCaptions.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("EnablePopOut", this.EnablePopOut.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("EnableOffline", this.EnableOffline.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("StartMuted", this.StartMuted.ToString().ToLower(CultureInfo.InvariantCulture));
            writer.WriteElementString("StretchNonSquarePixels", this.StretchNonSquarePixels.ToString());
            _Items.Serialize(writer);
            writer.WriteEndElement();
        }
        #endregion

    }
}

File: Chapter10/VB/VBProject_10/AdvancedMedia/1-AdvanceMedia/AdvMediaPlayer.Web/EncoderMedia/MediaPlayer/PlayList.vb
'  <copyright file="PlaylistCollection.cs" company="Microsoft">
'      Copyright © Microsoft Corporation. All rights reserved.
'  </copyright>
'  <summary>Implements the PlaylistCollection class</summary>
'  <author>Microsoft Expression Encoder Team</author>
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Windows
Imports System.Windows.Browser
Imports System.Windows.Media
Imports System.Threading
Imports System.Xml
Imports System.IO.IsolatedStorage
Imports System.IO
Imports System.Globalization
Imports System.Text

' Namespace ExpressionMediaPlayer

    ''' <summary>
    ''' exception for invalid playlist
    ''' </summary>
    Public Class InvalidPlaylistException
        Inherits Exception
        Public Sub New()
        End Sub '   New

        Public Sub New(message As String)

        {
        End Sub '   New

        Public Sub New(nodeFound As String, nodeExpected As String)

        {
        End Sub '   New
    End Class   '   InvalidPlaylistException

    '  Summary:
    '      Describes how content is resized to fill its allocated space -- taking into account non-square pixels
    public enum StretchNonSquarePixels
    {
        '  Summary:
        '      The content preserves its original size and frame aspect ratio
        NoStretch = 0,
        '
        '  Summary:
        '      The content is resized to fill the destination dimensions. The frame aspect ratio is preserved.
        StretchToFill = 1,
        '
        '  Summary:
        '      The content is resized to fill the destination dimensions. Original size and frame aspect ratio are ignore.
        StretchDistorted = 2,
    }

    <ScriptableType>
    Public Partial Class Playlist
        Inherits DependencyObject
    {
        #region PlaylistCore
        ''' <summary>
        ''' playlist options
        ''' </summary>
        Private m_autoLoad As Boolean
        Private m_autoPlay As Boolean
        Private m_autoRepeat As Boolean
        Private m_displayTimeCode As Boolean
        Private m_enableCachedComposition As Boolean
        Private m_enableCaptions As Boolean
        Private m_enableOffline As Boolean
        Private m_enablePopOut As Boolean
        Private m_startMuted As Boolean
        Private m_startWithPlaylistShowing As Boolean
        private StretchNonSquarePixels m_stretchMode
        Private m_background As Color

        ''' <summary>
        ''' The number of items in the playlist that are DRM protected
        ''' </summary>
        Private m_countOfDRMItems As Integer
        ''' <summary>
        ''' The number of items in the playlist that are Live Items
        ''' </summary>
        Private m_countOfLiveItems As Integer
        ''' <summary>
        ''' The number of items in the playlist that are adaptive Items
        ''' </summary>
        Private m_countOfAdaptiveItems As Integer

        ''' <summary>
        ''' list of playlist items
        ''' </summary>
        private PlaylistCollection _Items = New PlaylistCollection()
        ''' <summary>
        ''' playlist changed event
        ''' </summary>
        <ScriptableMember>
        public event RoutedEventHandler PlaylistChanged
        ''' <summary>
        ''' should playlist cue up
        ''' </summary>
        <Description("Automatically cue video when page is loaded"), DefaultValue(true)>
        Public Property AutoLoad() As Boolean
            Get

                return m_autoLoad
            End Get

            Set

                m_autoLoad = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' should playlist auto start?
        ''' </summary>
        <Description("Automatically start video when cued"), DefaultValue(true)>
        Public Property AutoPlay() As Boolean
            Get

                return m_autoPlay
            End Get

            Set

                m_autoPlay = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' should playlist auto repeat?
        ''' </summary>
        <Description("Automatically restart video when at end"), DefaultValue(false)>
        Public Property AutoRepeat() As Boolean
            Get

                return m_autoRepeat
            End Get

            Set

                m_autoRepeat = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' display timecodes
        ''' </summary>
        <Description("Display Timecode")>
        Public Property DisplayTimeCode() As Boolean
            Get

                return m_displayTimeCode
            End Get

            Set

                m_displayTimeCode = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' cached composition enabled
        ''' </summary>
        <Description("Enable Cached Composition"), DefaultValue(true)>
        Public Property EnableCachedComposition() As Boolean
            Get

                return m_enableCachedComposition
            End Get

            Set

                m_enableCachedComposition = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' captions enabled
        ''' </summary>
        <Description("Allow closed captions to show"), DefaultValue(true)>
        Public Property EnableCaptions() As Boolean
            Get

                return m_enableCaptions
            End Get

            Set

                m_enableCaptions = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' offlining enabled
        ''' </summary>
        <Description("Enable Player to be run offline"), DefaultValue(true)>
        Public Property EnableOffline() As Boolean
            Get

                return m_enableOffline  AndAlso  (m_countOfDRMItems < 1)  AndAlso  (m_countOfLiveItems < 1)
            End Get

            Set

                m_enableOffline = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' popout enabled
        ''' </summary>
        <Description("Enable Player popout"), DefaultValue(true)>
        Public Property EnablePopOut() As Boolean
            Get

                return m_enablePopOut
            End Get

            Set

                m_enablePopOut = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property


        ''' <summary>
        ''' HasAdaptiveItems
        ''' </summary>
        <Description("Indicated whether there are any Smooth Streaming items in the playlist")>
        Public ReadOnly Property HasAdaptiveItems() As Boolean
            Get

                return m_countOfAdaptiveItems > 0
            End Get
        End Property

        ''' <summary>
        ''' start in muted state
        ''' </summary>
        <Description("Mute player on start")>
        Public Property StartMuted() As Boolean
            Get

                return m_startMuted
            End Get

            Set

                m_startMuted = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property


        ''' <summary>
        ''' start with playlist visible
        ''' </summary>
        <Description("Show playlist on start")>
        Public Property StartWithPlaylistShowing() As Boolean
            Get

                return m_startWithPlaylistShowing
            End Get

            Set

                m_startWithPlaylistShowing = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' type of video stretch
        ''' </summary>
        <Description("Stretch Mode")>
        public StretchNonSquarePixels StretchNonSquarePixels
        {
            get
            {
                return m_stretchMode
            }
            set
            {
                m_stretchMode = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If

            }
        }
        ''' <summary>
        ''' color of background
        ''' </summary>
        Public Property Background() As Color
            Get

                return m_background
            End Get

            Set

                m_background = value

                If (PlaylistChanged  IsNot Nothing) Then
                    PlaylistChanged(this, Nothing)
                End If
            End Set
        End Property
        ''' <summary>
        ''' list of playlist items.
        ''' </summary>
        <Description("Playlist")>
        public PlaylistCollection Items
        {
            get
            {
                return _Items
            }
            set
            {
                _Items = value
            }
        }
        ''' <summary>
        ''' init structure
        ''' </summary>
        private Sub Init()

            AutoLoad = true
            AutoPlay = true
            AutoRepeat = false
            EnableCachedComposition = true
            EnableCaptions = true
            EnablePopOut = true
            EnableOffline = true
            m_countOfDRMItems = 1
            m_countOfLiveItems = 1
            m_countOfAdaptiveItems = 0
        End Sub '   Init
        ''' <summary>
        ''' construct a playlistitem, provided for scripting.
        ''' </summary>
        ''' <returns>new playlist item</returns>
        public PlaylistItem CreateNewPlaylistItem()
        {
            return New PlaylistItem(Items)
        }
        ''' <summary>
        ''' playlist constructor
        ''' </summary>
        Public Sub New()
            Me.StretchNonSquarePixels = StretchNonSquarePixels.NoStretch
            _Items = New PlaylistCollection()
            AddHandler _Items.CollectionChanged, AddressOf System.Collections.Specialized.NotifyCollectionChangedEventHandler(Items_CollectionChanged)
            Init()
        End Sub '   New
        ''' <summary>
        ''' inform playlist changed if the items collection changed
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        private Sub Items_CollectionChanged(sender As Object, e As System.Collections.Specialized.NotifyCollectionChangedEventArgs)


            If (PlaylistChanged <> Nothing) Then
                PlaylistChanged(this, Nothing)
            End If

            m_countOfDRMItems = 0
            m_countOfLiveItems = 0
            m_countOfAdaptiveItems = 0

            For Each var item in Me.Items


                If (item.DRM) Then
                    m_countOfDRMItems += 1
                End If


                If (item.IsAdaptiveStreaming) Then
                    m_countOfAdaptiveItems += 1

                    Dim source As String  = item.MediaSource.OriginalString.ToLower()

                    if (source.Contains(@".isml/manifest"))
                    {
                        m_countOfLiveItems += 1
                    }
                else

                    If (item.MediaSource.IsAbsoluteUri) Then

                        Dim scheme As String  = item.MediaSource.Scheme.ToLower()


                        If (scheme = ("mms:")) Then
                            m_countOfLiveItems += 1
                        End If
                    End If
                End If
            Next    '   var
        End Sub '   Items_CollectionChanged
        ''' <summary>
        ''' create playlist from XML string
        ''' </summary>
        ''' <param name="playlistXml">XML string</param>
        ''' <returns>playlist</returns>
        Public ReadOnly Property Playlist() As
        ''' <summary>
        ''' parse playlist from XML
        ''' </summary>
        ''' <param name="playlistXml"></param>
        public Sub ParseXml(playlistXml As String)

            UTF8Encoding enc = New UTF8Encoding()
            using (MemoryStream ms = New MemoryStream(enc.GetBytes(playlistXml)))
            {
                XmlReaderSettings xmlrs = New XmlReaderSettings()
                xmlrs.IgnoreComments = true
                xmlrs.IgnoreWhitespace = true

                Dim reader As XmlReader  = XmlReader.Create(ms, xmlrs)

                Deserialize(reader)
            }

            If (PlaylistChanged  IsNot Nothing) Then
                PlaylistChanged(this, Nothing)
            End If
        End Sub '   ParseXml
        #End Region

        #region Serialization
        ''' <summary>
        ''' top level XML node for this class
        ''' </summary>
        internal const string xmlNode = "Playlist"
        ''' <summary>
        ''' deserialise this object
        ''' </summary>
        ''' <param name="reader">XmlReader to deserialize from</param>
        ''' <returns>this</returns>
        internal Sub Deserialize(reader As XmlReader)


            If ( Not reader.IsStartElement(Playlist.xmlNode)) Then
                throw New InvalidPlaylistException()
            End If

            Init()
            reader.Read()
              Playlist.xmlNode  AndAlso  reader.NodeType = XmlNodeType.EndElement))
              While ( Not (reader.Name = XmlNodeType.EndElement))


                If (reader.IsStartElement("AutoLoad")) Then
                    Me.AutoLoad = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("AutoPlay"))
                    Me.AutoPlay = reader.ReadElementContentAsBoolean()
                End If
                else if (reader.IsStartElement("AutoRepeat"))
                    Me.AutoRepeat = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("DisplayTimeCode"))
                    Me.DisplayTimeCode = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("EnableCachedComposition"))
                    Me.EnableCachedComposition = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("EnableCaptions"))
                    Me.EnableCaptions = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("EnablePopOut"))
                    Me.EnablePopOut = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("EnableOffline"))
                    Me.EnableOffline = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("StartMuted"))
                    Me.StartMuted = reader.ReadElementContentAsBoolean()
                else if (reader.IsStartElement("StartWithPlaylistShowing"))
                {
                    Me.StartWithPlaylistShowing = reader.ReadElementContentAsBoolean()
                }
                else if (reader.IsStartElement("StretchMode"))
                {

                    Dim tmp As Stretch  = (Stretch)Enum.Parse(GetType(Stretch), reader.ReadElementContentAsString(), false)

                    switch(tmp)
                    {
                        case Stretch.None:
                            Me.StretchNonSquarePixels = StretchNonSquarePixels.NoStretch

                        case Stretch.Uniform:
                        case Stretch.UniformToFill:
                            Me.StretchNonSquarePixels = StretchNonSquarePixels.StretchToFill

                        case Stretch.Fill:
                            Me.StretchNonSquarePixels = StretchNonSquarePixels.StretchDistorted

                    }
                }
                else if (reader.IsStartElement("StretchNonSquarePixels"))
                {
                    StretchNonSquarePixels tmp = (StretchNonSquarePixels)Enum.Parse(GetType(StretchNonSquarePixels), reader.ReadElementContentAsString(), false)
                    Me.StretchNonSquarePixels = tmp
                }
                else if (reader.IsStartElement(PlaylistCollection.xmlNode))
                {
                    Me.Items.Clear()
                    Me.Items.Deserialize(reader)
                }
                else if (reader.IsStartElement())
                    throw New InvalidPlaylistException(reader.Name, Playlist.xmlNode)
                else if ( Not reader.Read())
            End While   '
        End Sub '   Deserialize
        ''' <summary>
        ''' serialize this object
        ''' </summary>
        ''' <param name="writer">XmlWriter</param>
        internal Sub Serialize(writer As XmlWriter)

            writer.WriteStartElement(Playlist.xmlNode)
            writer.WriteElementString("AutoLoad", Me.AutoLoad.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("AutoPlay", Me.AutoPlay.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("AutoRepeat", Me.AutoRepeat.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("DisplayTimeCode", Me.DisplayTimeCode.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("EnableCachedComposition", Me.EnableCachedComposition.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("EnableCaptions", Me.EnableCaptions.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("EnablePopOut", Me.EnablePopOut.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("EnableOffline", Me.EnableOffline.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("StartMuted", Me.StartMuted.ToString().ToLower(CultureInfo.InvariantCulture))
            writer.WriteElementString("StretchNonSquarePixels", Me.StretchNonSquarePixels.ToString())
            _Items.Serialize(writer)
            writer.WriteEndElement()
        End Sub '   Serialize

        #End Region
    End Class   '   Playlist
' End Namespace   '   ExpressionMediaPlayer
' ..\Project_10\AdvancedMedia\1-AdvanceMedia\AdvMediaPlayer.Web\EncoderMedia\MediaPlayer\PlayList.cs
